### PR TITLE
Add `CertificateChainVerifier` implementation

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
A `CertificateChainVerifier` implementation is now available using
mbedtls. This implementation is behind the `x509` feature flag.

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #131** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/131" target="_blank">Graphite</a> 
    * **PR #132** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/132" target="_blank">Graphite</a> 
      * **PR #135** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/135" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/135?utm_source=stack-comment).

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #131** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/131" target="_blank">Graphite</a> 
    * **PR #132** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/132" target="_blank">Graphite</a> 
      * **PR #135** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/135" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/135?utm_source=stack-comment).

File: verifier/src/x509.rs [hunk 1]
```diff
@@ -29,6 +35,53 @@ impl From<mbedtls::Error> for Error {
...
+                CertificateChainVerifierError::SignatureVerification
+            }
+            _ => CertificateChainVerifierError::GeneralCertificateError,
+        }
+    }
+}
+
+/// A certificate chain verifier that uses MbedTls as the backend
+#[derive(Debug)]
+pub struct X509CertificateChainVerifier {
```
Comment (nick-mobilecoin) [inline verifier/src/x509.rs:59]: not fond of this name, seems too wordy and possibly ambiguous with the `x509_cert` crate. Maybe mbedtls should be in the name, but that feels like it's exposing the backend that is meant to be hidden, other than mbedtls seems to be a PIA for the different SDK platforms.

File: verifier/src/x509.rs [hunk 2]
```diff
@@ -1,18 +1,22 @@
 // Copyright (c) 2023 The MobileCoin Foundation
 
 extern crate alloc;
-use alloc::string::{String, ToString};
+use alloc::string::String;
```
Comment (eranrund) [inline verifier/src/x509.rs:4]: Nit that's not directly related to this PR but I'm surprised imports are not being grouped here, I thought we standardized on grouping them (which might require a tweak to `rustfmt.toml`).

File: verifier/src/x509.rs (hunk 1 above)
Comment (eranrund) [reply verifier/src/x509.rs:59]: With the risk of bikeshedding, I think I would've chosen `MbedtlsCertificateChainVerifier` (named a feature that brings in `mbedtls` in the crate), and the file probably `mbedtls.rs` or something like that. The thinking behind that the trait `CertificateChainVerifier` already implies these are X509 certs, given that one of the arguments of its only method accepts a list of 509 certificates. So it doesn't seem like there's a world in which `CertificateChainVerifier` verifies any other kind of certificate, so the implementation is really just selecting which backend one wants to use. And as such I would've named the backend implementation that uses `mbedtls` `MbedtlsCertificateChainVerifier`.

File: verifier/src/x509.rs [hunk 3]
```diff
@@ -29,6 +35,53 @@ impl From<mbedtls::Error> for Error {
...
+
+impl X509CertificateChainVerifier {
+    /// Create a new instance
+    pub fn new(trust_anchor: TrustAnchor) -> Self {
+        Self { trust_anchor }
+    }
+}
+
+impl CertificateChainVerifier for X509CertificateChainVerifier {
+    // Note: `_time` is ignored because MbedTls will call out to a system timer.
```
Comment (eranrund) [inline verifier/src/x509.rs:71]: Boo, not an issue for this PR to address, but that's a bummer since I think this means we can't run this code inside an enclave. I am not sure that's something we would've needed, I don't have the full picture on what enclave-to-enclave attestation looks like.

File: verifier/src/x509.rs [hunk 4]
```diff
@@ -333,35 +311,56 @@ mod test {
...
-        let crl = CertificateRevocationList::try_from_pem([ROOT_CRL, PROCESSOR_CRL])
-            .expect("failed to parse CRL");
-        assert!(cert_chain.verify(&trust_anchor, crl).is_ok());
+        let crls = [ROOT_CRL, PROCESSOR_CRL]
+            .iter()
+            .map(|crl| CertificateList::from_der(crl).expect("failed to parse CRL"))
+            .collect::<Vec<_>>();
+        let verifier = X509CertificateChainVerifier::new(trust_anchor);
+        assert!(verifier
+            .verify_certificate_chain(chain.iter(), crls.iter(), time)
```
Comment (eranrund) [inline verifier/src/x509.rs:327]: Does the time here actually matters?

File: verifier/src/x509.rs (hunk 2 above)
Comment (nick-mobilecoin) [reply verifier/src/x509.rs:4]: It's not stable :(, https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#imports_granularity
the mobilecoin repo uses nightly while we had kept this and the SGX repo on stable, with a ci build for nightly.

File: verifier/src/x509.rs (hunk 1 above)
Comment (nick-mobilecoin) [reply verifier/src/x509.rs:59]: I will update to that, I wasn't really fond of the x509 naming, and I agree "certificate chain" => x509

File: verifier/src/x509.rs (hunk 4 above)
Comment (nick-mobilecoin) [reply verifier/src/x509.rs:327]: nope, i missed this when I did some refactoring cleanups to use a constant `DateTime` value.  Will fix to jive with the other test cases.

File: verifier/src/x509.rs (hunk 3 above)
Comment (nick-mobilecoin) [reply verifier/src/x509.rs:71]: > since I think this means we can't run this code inside an enclave.

The time in enclaves has to come from untrusted, so it's not trusted anyway :(.
However mbedtls is currently used in the enclaves, it doesn't use time, maybe I need to update the comment:
- https://github.com/mobilecoinfoundation/rust-mbedtls/blob/6d8fe323a3292f87a6bce4b35963d47139a583f9/mbedtls-sys/vendor/library/x509.c#L984 When `MBEDTLS_HAVE_TIME_DATE` is not defined.
- The default value of `MBEDTLS_HAVE_TIME_DATE`, https://github.com/mobilecoinfoundation/rust-mbedtls/blob/6d8fe323a3292f87a6bce4b35963d47139a583f9/mbedtls-sys/build/config.rs#L74
- An option to enable time/date, https://github.com/mobilecoinfoundation/rust-mbedtls/blob/6d8fe323a3292f87a6bce4b35963d47139a583f9/mbedtls-sys/build/config.rs#L420
  - It doesn't appear to be used for `"time"`, https://github.com/mobilecoinfoundation/rust-mbedtls/blob/6d8fe323a3292f87a6bce4b35963d47139a583f9/mbedtls-sys/build/features.rs#L22

File: verifier/src/x509.rs (hunk 4 above)
Comment (nick-mobilecoin) [reply verifier/src/x509.rs]: updated now

File: verifier/src/x509.rs (hunk 1 above)
Comment (nick-mobilecoin) [reply verifier/src/x509.rs]: updated now

File: verifier/src/x509.rs (hunk 2 above)
Comment (nick-mobilecoin) [reply verifier/src/x509.rs]: I fixed all of them, for now...
We may need to revisit if we want to start using nightly for these repos.

File: verifier/src/x509.rs (hunk 3 above)
Comment (nick-mobilecoin) [reply verifier/src/x509.rs]: Updated the comment to say that time is ignored all the way down for mbedtls:(.

**Review by eranrund — APPROVED** [review]

File: verifier/src/x509.rs (hunk 2 above)
Comment (eranrund) [reply verifier/src/x509.rs]: Ha, shame on them. Seems like such a simple thing to stabilize 🤷‍♂️

File: verifier/src/x509.rs (hunk 3 above)
Comment (eranrund) [reply verifier/src/x509.rs]: Makes sense, thanks for the detailed walk-through for the mbedtls implementation. Comment looks good.
